### PR TITLE
Add FXIOS-10031 [Menu] Redux for menu detail view controller

### PR DESCRIPTION
**Pull request by adudenamedruby**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-10031)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/22004)

## :bulb: Description
This adds Redux to the Menu details screen. Due to blockers, this refactor has taken some extra time. As such, I'm gonna ask for review to ignore a few things:
- `MenuConfigurationUtility`: For now, please ignore optionals in initializers. These will be removed in the PR where we add the remaining actions to the menu and finalize the actual utility.
- No store tests or tests for this PR: the PR was already getting big and I've had to rebase and fix so many conflicts over the last two weeks for this thing. A ticket has been created specifically to be the next PR (#22363) to go back in and add the missing tests.
- A few Redux anti-patterns: my discovery of the error of `return state` in guard statements in Redux has caused me to fix the immediate issues, but few of these `return state` things (that have not yet been causing problems) remain. I'm leaving them in there for now, until either a) Orla has come up with pattern for a general a solution for this, or b) I'm gonna make a helper function and return that in a PR meant to address this issue, at least for menu work as a whole.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

**Comment by adudenamedruby** [comment]
added @OrlaM as per my slack comment to her re a pattern for the Redux recursion problems

File: firefox-ios/Client/Frontend/Browser/MainMenu/Views/MainMenuDetailsViewController.swift [hunk 1]
```diff
@@ -121,12 +110,87 @@ class MainMenuDetailViewController: UIViewController,
...
+                actionType: ScreenActionType.closeScreen,
+                screen: .mainMenuDetails
+            )
+        )
+    }
+
+    func newState(state: MainMenuDetailsState) {
+        submenuState = state
+
+        refreshContent()
```
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Browser/MainMenu/Views/MainMenuDetailsViewController.swift:170]: Do we want to call this method only on a certain state update? It seems this reloads the entire table view, so just wanted to make sure we only want to reload when needed.

**Review by cyndichin — APPROVED** [review]
one minor comment, but otherwise looks good!

**Comment by cyndichin** [comment]
I missed this, but seems like bitrise is erroring out on 

>   /Users/vagrant/git/firefox-ios/Client/Frontend/Browser/MainMenu/Views/MainMenuDetailsViewController.swift:172:25: value of type 'MainMenuDetailsState' has no member 'shouldGoBackToMainMenu'
        if submenuState.shouldGoBackToMainMenu {

File: firefox-ios/Client/Frontend/Browser/MainMenu/Views/MainMenuDetailsViewController.swift (hunk 1 above)
Comment (adudenamedruby) [reply firefox-ios/Client/Frontend/Browser/MainMenu/Views/MainMenuDetailsViewController.swift]: I see what you're saying. I think it's small enough that it kinda doesn't matter - and as far as I understand, we shouldn't be optimizing views updating themselves with the same information. I'll ask in the channel, just to get clarification, and then I'll update accordingly.

Also, I meant to have this at the end of the function, so if you're taking an available action, you're not actually refreshing one last time for no reason. I'll make that change.

**Pull request merged**